walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,83 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Return all times if there are no attendees
+    if (request.getAttendees().isEmpty() && request.getOptionalAttendees().isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    } 
+
+    ArrayList<String> mandatoryAttendees = new ArrayList<String>(request.getAttendees());
+    ArrayList<String> allAttendees = new ArrayList<String>(mandatoryAttendees);
+    allAttendees.addAll(request.getOptionalAttendees());
+
+    ArrayList<TimeRange> timeRanges = new ArrayList<TimeRange>(findTimeRanges(allAttendees, events, request));
+
+    // If no time ranges are available, exclude optional attendees and try again
+    if (!timeRanges.isEmpty()) {
+      return timeRanges;
+    } else {
+      timeRanges = new ArrayList<TimeRange>(findTimeRanges(mandatoryAttendees, events, request));
+    }
+
+    // If the request has no attendees after optional attendees are excluded, return an empty collection.
+    if (!mandatoryAttendees.isEmpty()) {
+      return timeRanges;
+    } else {
+      return new ArrayList<TimeRange>();
+    }
+  }
+
+  public Collection<TimeRange> findTimeRanges(ArrayList<String> attendees, Collection<Event> events, MeetingRequest request) {
```

Comment:
looks like MeetingRequest is only used for getting duration. we can just pass in duration instead of the  MeetingRequest object

Reply:
done.